### DISCUSSION TITLE
POC: use fuse-overlayfs to use your system firefox

### DISCUSSION THREAD

**Discussion by akdor1154** [filips123/PWAsForFirefox#67] [Ideas]
Hey,
For starters, thanks heaps for this project!

I've been stuffing around with trying to use overlayfs to use my system's firefox runtime instead of having FirefoxPWA download a new one for me. This nearly works perfectly ;)
Launching from CLI actually works fine:
```sh
cd ~/.local/config/firefoxpwa
rm -rf runtime # or mv to old_, etc.
mkdir upper work runtime
# set up `runtime` with system firefox as the lower directory
fuse-overlayfs -o lowerdir=/usr/lib/firefox,upperdir=upper,workdir=$(pwd)/work,squash_to_uid=$(id -n) $(pwd)/runtime
# now launch
firefoxpwa site launch 01FJFV1VC9YCAVP2KVGQ86VSXW # ms teams, ugh
```

This works great - runtime.patch works fine and I can see all the modifications in the /upper directory, and the PWA works fine!

The one issue I have yet to figure out is that launching from the browser doesn't work. The subprocess `firefox` does seem to start with the correct arguments, but it doesn't show any window. Maybe there are env vars that are persisting from the parent process? I'll look more into this, I can't imagine this is insurmountable given cli launch works fine.

Just sharing in case it sparks interest! This is on Ubuntu 21.10 for reference, but I think fuse-overlayfs has worked ok for a while now.

**Comment by akdor1154** [comment]
Sort of found why it was breaking when launching from the extension:
if MOZ_ENABLE_WAYLAND is set, then the pwa process starts but seems to hang early in startup - it never shows a window but I can see it in `ps`.
I tried pulling out various env vars that were set by the initial browser (a few from ubuntu/debian's startup script), to no avail.
I turned on logging to a file to see what was going on, and bang, there is my working PWA!
So it turns out for some bizarre reason, that if  MOZ_ENABLE_WAYLAND=1, you also need to set MOZ_LOG_FILE to a non-empty value or it will hang on startup. :shrug:

**Comment by filips123** [comment]
This is interesting. I unfortunately don't know why Firefox requires `MOZ_LOG_FILE`, but it's nice you managed to figure this out.

Because fuse-overlayfs only works on Linux and has this problem with env vars, I will probably just add manual instructions to the repo wiki or extension. In the future, I might add an additional option to set up fuse-overlayfs instead of downloading runtime.

Also, if you find why this problem happened or how to properly fix it, please let me know.

**Comment by filips123** [comment]
As I'm not really experienced in `fuse-overlayfs`, I have another question. Is it possible to automatically set up `fuse-overlayfs` when the system starts/user logs in/when needed? Apparently, it should be possible by adding some entries into `/etc/fstab`, but I'm not sure if this will work because it needs to be mounted into a user-specific directory, and for each user.

**Reply by akdor1154** [reply]
Yep it would be technically possible, but I actually wouldn't go down that path if looking to make this smoother - in theory it's possible to just do at runtime by spawning `fuse-overlayfs` - this doesn't require root and I think works OOTB on at least Ubuntu 21.04 onwards (as long as fuse-overlayfs is installed).
So what I had in the back of my mind was playing with FirefoxPWA itself to try and run this.

Of course while I'm being lazy and haven't played with the above yet, I have /etc/fstab (actually systemd mount units) set up like in your message and it works fine :)

**Comment by filips123** [comment]
I added instructions for using FUSE OverlayFS for the runtime. With the provided systemd user service, the runtime will be automatically linked when the user logs in, without the need to additional configuration or root access. See #214 for more details.
